base/loading.jl

```diff
@@ -1411,6 +1413,20 @@ function srctext_files(f::IO, srctextpos::Int64)
     return files
 end
 
+function get_preferences_hash(uuid::UUID, cache::TOMLCache = TOMLCache())
+    # check that project preferences match by first loading the Project.toml
+    active_project_file = Base.active_project()
+    if isfile(active_project_file)
+        preferences = get(parsed_toml(cache, active_project_file), "preferences", Dict{String,Any}())
```

Comment:
Sorry if I missed earlier discussion on this. But wouldn't this behave a bit confusingly when you stack environments?

Suppose I have `LOAD_PATH = ["@", "@v#.#", "@stdlib"]` and some preferences for SomeREPLUtilities.jl in `~/.julia/environments/v1.6/Project.toml`. If I activate `$PWD/Project.toml` and then evaluate `using SomeREPLUtilities`, does it mean the preferences in `~/.julia/environments/v1.6/Project.toml` would be ignored?  How about looking at all project files in `Base.load_path()`?

Reply:
It should *only* check the project file where `SomeREPLUtilities` is found I guess. If you need local preferences for that package you should probably add it to the local project.

Reply:
Ah, yes, good point. I agree.

Reply:
It's purposeful that it only checks the top-level project file.  We decided against any kind of merging for a few reasons:

1) Reproducibility; if a Project.toml somewhere high up in a chain is effecting things, then simply saving the `Project.toml` and `Manifest.toml` that you normally do would not be enough to guarantee identical results later on; we want all the configuration stored in one place.

2) Performance; we don't want to have to touch a bunch of files.  This can be ameliorated with caches, of course, but it's nice that we only have to read in one file.

3) Merging complexity; since we're storing full-on `Dict`'s here, what happens if a higher-up and lower-down Projects have conflicts around key mappings?

I'm torn on adding preferences for packages not in the project; it makes sense to disallow it, but it's also harmless.

Reply:
I favor disallowing the inclusion of preferences for packages not in the project simply because allowing it allows a[two] non-involved package[s] to carry preferences that may conflict were that[those] package[s] to appear in the project at a future time. It works against clear communication in collaboration.

Reply:
1. But if you use a package not in the current project you can't reproduce anyway so saving preferences for a non-existing package does not really help anything?
2. I don't think this will have any impact at all, we still go to that same file to find the UUID of the package (and with the new TOML caching we  have alraedy parsed that file).
3. That is a good point, you mean if some pakcage X has config in multiple places? I would say just use the first one then.

Reply:
@staticfloat Thanks for explaining the context. Just to be explicit, I'm not suggesting to look into package's project file (in the sense `joinpath(pkgdir(MyPkg), "Project.toml")`) and also I'm not suggesting to do any "deep merge." Concretely, I'm suggesting something like

```julia
function get_preferences_hash(uuid::UUID, cache::TOMLCache = TOMLCache())
    # check that project preferences match by first loading the Project.toml
    for path in Base.load_path()
        if isfile(path)
            project = parsed_toml(cache, path)
            # `uuid` is a dependency:
            any(values(get(project, "deps", Dict{String,Any}()))) do uuid
                tryparse(UUID, uuid) == uuid
            end || continue
            # Now we are looking at correct Project.toml; let's load `preferences`
            preferences = get(project, "preferences", Dict{String,Any}())
            if haskey(preferences, string(uuid))
                return UInt64(hash(preferences[string(uuid)]))
            end
            break  # should we?
        end
    end
    return UInt64(hash(Dict{String,Any}()))
end
```

It'd only look up just a few files in `Base.load_path()` (typically one or two, I think). So, I'm not sure if there is too much performance overhead.

I think my implementation addresses the points raised by @fredrikekre.

> 2\. (and with the new TOML caching we  have alraedy parsed that file)

@fredrikekre I guess caching is not useful during precompilation since it'd happen in a separate process? Though I agree that parsing a few files instead of just one file does not seem to be a big impact on the performance.

Reply:
> I think my implementation addresses the points raised by @fredrikekre.

Yea, that is what I meant -- find the file where the package would be `using`d from.

> I guess caching is not useful during precompilation since it'd happen in a separate process?

True, I was only thinking of performance for getting/setting prefs.

Reply:
It might be fine to allow preferences for any package as long as you take preferences from the environment in whose manifest the package first appears. The reason that's ok is because that guarantees that preferences later in the load path can only affect packages that nothing earlier in the load path depends on. To spell that out, if A is a top-level package in the first environment in the load path and it depends on B and B has no preferences set in the first environment (because, e.g it is not a top-level package), then we need to load B with no preferences set. If, on the other hand, C is in the second environment and not depended on by anything in the first environment, then we can apply preferences for it from the second environment.

Reply:
Another point: If we *only* look in the top project file, then you need to configure the package in _every project you use it in_. E.g. if you have Revise in the global env, for every project you want to use Revise with you have to configure it again.

Reply:
> To spell that out, if A is a top-level package in the first environment in the load path and it depends on B and B has no preferences set in the first environment (because, e.g it is not a top-level package), then we need to load B with no preferences set. If, on the other hand, C is in the second environment and not depended on by anything in the first environment, then we can apply preferences for it from the second environment.

This is consistent with the idea that you load preferences from the project where the package is `using`d from, right? Since `B` would be found in the top manifest.

Reply:
> Concretely, I'm suggesting something like

Ahhh, I understand what you're saying now.  Thanks, that was really helpful.  Yes, I think such an approach makes sense!  And I believe that Stefan's point is that the `break` in your `for` loop should exist, correct?

Reply:
IIUC, there may be a subtle difference between what I suggested and what Stefan suggested. Suppose, as in Stefan's setting, I have:

* Packages: A, B
* Dependencies: A depends on B
* Environments (`LOAD_PATH`):
  * X: It contains A as a dependency (i.e., `deps` has `A = "u-u-i-d"`).
  * Y: It contains B as a dependency but A is not in the manifest. It has preferences for B.

With my suggestion, `using A` would load B with the preference stored in Y.  I cannot tell if B's preference in Y would be used in Stefan's approach.

The reason why I like my approach is that I think `using B` in the stacked environment X+Y should behaves as close as possible to single environment Y if B is installed explicitly only in Y (although it's not possible to do it always). Typically, I think Y is `@v#.#` and X is `@`. For example, if you have Revise.jl in `@#.#` and it has some preferences, you'd want to use the preferences recorded in `@#.#` even if you have a package that depends on Revise.jl in `@`.

(In the above setting `break`-or-not does not matter. So I'm not sure how it interacts with Stefan's suggestion.)

Reply:
Ah, in that case, I think Stefan's approach makes more sense.

The reason being that we try hard not to allow things "lower down" in the stack to change the behavior of things at the top of the stack.  For instance, if I execute something where only `X` is in my load path, I will get the "no preferences" behavior, but if I execute the same project but with `Y` inserted underneath `X` in the stack, suddenly the behavior changes.  I think this is generally undesirable.

In the example of `Revise`, I think it makes sense to load the preferences for `Revise` from your global environment if `Revise` is not installed in your local package, but if it is, I do not think we should use the global environment's values.  One potential reason why is that if I were to set some custom options in my global environment, it then becomes difficult to have a "clean" environment where I get only the defaults without setting those defaults explicitly.

Reply:
@KristofferC if I wanted to walk the `load_path()` and test at each point as to whether or not a particular UUID is contained within the manifest there, what's the best way to do that?  I don't want to check the `deps` section of the `Project.toml` because that only contains direct dependencies, I want to check everything in the Manifest.  Is there an already-loaded cache of that somewhere in `Base`, or should I just search for `Manifest.toml` files sitting within/next to the elements of `load_path()` and parse them directly?

Reply:
Hmm... My intuition is still that preference for B in Y should be loaded even if B is in X. This is because the fact that A depends on B is an implementation detail. For example, A may stop depending on B when changing from 1.1 to 1.2. Then `Pkg.up` on X (that removes B from X) suddenly changes the behavior of `using B` even when Y is completely untouched and the user doesn't know if B was in X. I think this behavior is counter-intuitive.

> it then becomes difficult to have a "clean" environment where I get only the defaults without setting those defaults explicitly

One solution may be to introduce a "dummy" environment (say) `@nopref` that "resets" preferences.  For example, `LOAD_PATH = ["@", "@nopref", "@#.#", "@stdlib"]` would only load preferences from `@`.  It's an additional complexity w.r.t implementation but I think it's much easier to control preferences this way.  Even with the current implementation, it's a bit hard to fully predict what preferences are used unless you know all the entries in the manifest files.

Reply:
> This is because the fact that A depends on B is an implementation detail.

That implementation detail has big consequences though; think about it from a versioning perspective.  If `A` restricts package `B` to `v1.1` in environment `Y`, then even though environment `X` includes `B` as a direct dependency and would normally allow it to be any `v1.X` version, it will be restricted to `v1.1`.  In this manner, the behavior of `using B` is already changing based on transitive dependencies.

I'm not 100% convinced I'm right, but I'm still mostly convinced that apply preferences based on a UUID's presence in `Manifest.toml` is better than doing this based on presence in `Project.toml` (e.g. only for direct dependencies).  @StefanKarpinski are you still convinced?

Reply:
I agree with @tkf: If you need preferences on `B` in `Y` you can just add `B` such that it exists in the `Project.toml` file (https://github.com/JuliaLang/julia/pull/37595#discussion_r488979943).

I agree that it will be difficult to get a clean env though. But perhaps it is simple enough to play around with the LOAD_PATH then? I for one quite often use `export JULIA_LOAD_PATH=$(mktemp -d)` when I need to play around with a clean env.

Reply:
However, it does become a bit messy; when you load `A` and thus `B` is loaded implicitly, should you really go look for deps for `B` lower in the load path in a project that might have a different version of `B`?

Reply:
> In this manner, the behavior of `using B` is already changing based on transitive dependencies.

Yeah, I agree that's an issue. But isn't it an issue somewhat orthogonal to preferences? I think it'd be better to only support the case where the manifests are compatible and explicitly state that otherwise, the behavior is undefined. I think it's better to solve the issue with incompatible manifests with something like #32906.

But I can see that combining two manifests can change the version of a package in a manner incompatible with the stored preferences.

_If_ we can change how the package resolution works in a stacked environment, how about

1. When looking for a package to load, first try stacked *`Project.toml`* files. This way, the package version explicitly added by the user takes precedence over the indirect dependencies in the environments above.
2. Allow preferences to be set only when the package is added by the user (i.e., in `deps`).
3. For a given package, use the first `Project.toml` that contains the package as a dependency. If there are no preferences, use the default. (This is my implementation in https://github.com/JuliaLang/julia/pull/37595#discussion_r489018264)

I think the combination of these three ingredients ensure that preferences and package version are compatible while allowing the scenario I described in https://github.com/JuliaLang/julia/pull/37595#discussion_r489077392

Reply:
Any update on this? I think falling back to the lower environment stack is a key feature. I think one of the important use-cases for CompilePreferences.jl is to capture machine-specific configuration such as system libraries and hardware properties. For this, `~/.julia/environments/v1.6/Project.toml` sounds like a good place to store such information. It doesn't make sense to share `Project.toml` with machine-specific values.

I think I need this for PyPreferences.jl https://github.com/JuliaPy/PyCall.jl/pull/835. If it is decided that it is not appropriate to add this to CompilePreferences.jl, I think I'll create SharedPreferences.jl to handle this use-case (e.g., by storing the preferences in `~/.julia/sharedpreferences/$UUID/v$major.toml` and call `include_dependency` on it).  I'll look up per-project preferences first via CompilePreferences.jl and then fallback to SharedPreferences.jl. The fact that it is implementable outside `Base` *could* mean that not adding this in `Base`/CompilePreferences.jl is a good idea.  But having many ways to store preferences may be confusing.  I'm not sure what's the best approach here.

Reply:
I opened an issue for this discussion https://github.com/JuliaLang/julia/issues/37791